diff --git a/Makefile.am b/Makefile.am
index 845e89d..5f4d609 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -41,7 +41,8 @@ stamp-generator: icd_generator.rb
 
 # ignore the warning in OpenCL headers when using old interface
 libOpenCL_la_CFLAGS= $(NO_DEPRECATED_DECLARATIONS_FLAGS) \
-	$(AM_CFLAGS) $(PTHREAD_CFLAGS)
+	$(AM_CFLAGS) $(PTHREAD_CFLAGS) \
+	-DETC_OPENCL_VENDORS=\"@OCL_ICD_VENDORDIR@\"
 
 libOpenCL_la_SOURCES = ocl_icd_loader.c ocl_icd_loader.h ocl_icd_debug.h
 nodist_libOpenCL_la_SOURCES = ocl_icd_loader_gen.c ocl_icd.h
diff --git a/configure.ac b/configure.ac
index 264d6e0..bd703dd 100644
--- a/configure.ac
+++ b/configure.ac
@@ -145,6 +145,14 @@ AC_ARG_ENABLE([update-database],
   [update_database=no])
 AM_CONDITIONAL([UPDATE_DATABASE], [test x"$update_database" != xno])
 
+# --enable-custom-vendordir
+AC_ARG_ENABLE([custom-vendordir],
+  [AS_HELP_STRING([--enable-custom-vendordir],
+                  [use the given directory instead of /etc/OpenCL/vendors to look for .icd files])],
+  [OCL_ICD_VENDORDIR=$enableval],
+  [OCL_ICD_VENDORDIR=/etc/OpenCL/vendors])
+AC_SUBST([OCL_ICD_VENDORDIR])
+
 # always use versionned symbols (check required for MacOSX)
 AM_CONDITIONAL([USE_MAP], [true])
 
diff --git a/ocl_icd_loader.c b/ocl_icd_loader.c
index 831f2a7..06c5925 100644
--- a/ocl_icd_loader.c
+++ b/ocl_icd_loader.c
@@ -49,8 +49,6 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #define DEBUG_OCL_ICD_PROVIDE_DUMP_FIELD
 #include "ocl_icd_debug.h"
 
-#define ETC_OPENCL_VENDORS "/etc/OpenCL/vendors"
-
 int debug_ocl_icd_mask=0;
 
 typedef __typeof__(clGetPlatformInfo) *clGetPlatformInfo_fn;
